ios-training-bjung/Common/Mapper.swift

```diff
@@ -0,0 +1,48 @@
+//
+//  Mapper.swift
+//  ios-training-bjung
+//
+//  Created by 鄭 普勤 on 2023/12/12.
+//
+
+import Foundation
+
+final class Mapper {
+    static let shared = Mapper()
+    private init() {}
+    
+    private static let dateFormatter: DateFormatter = {
+        let dateFormatter = DateFormatter()
+        dateFormatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ssZZZZZ"
+        return dateFormatter
+    }()
+    
+    private let encoder: JSONEncoder = {
+        let encoder = JSONEncoder()
+        encoder.keyEncodingStrategy = .convertToSnakeCase
+        encoder.dateEncodingStrategy = .formatted(dateFormatter)
+        return encoder
+    }()
+    
+    private let decoder: JSONDecoder = {
+        let decoder = JSONDecoder()
+        decoder.keyDecodingStrategy = .convertFromSnakeCase
+        decoder.dateDecodingStrategy = .formatted(dateFormatter)
+        return decoder
+    }()
+    
+    func encodeWeatherRequest(request: WeatherRequest) throws -> String {
+        let jsonData = try encoder.encode(request)
+        guard let result = String(data: jsonData, encoding: .utf8) else { throw AppError.jsonEncodeError }
```

Comment:
```suggestion
        guard let jsonData = try? encoder.encode(request), let result = String(data: jsonData, encoding: .utf8) else { throw AppError.jsonEncodeError }
```

現時点では `WeatherRequest` に浮動小数点数型のプロパティーがないため問題になることはなさそうですけれど、それがあったときには `encoder.encode` が `EncodingError` を発生させる可能性が考えられます。もしかすると将来、それ以外の可能性も増えるかも知れません。

現状のコードで仮に `EncodingError` などが発生した場合、`HomePresenter.loadWeatherData()` のエラー処理で「原因不明のエラー」と判定されますが、それは意図した動作ですか？

仮に `Mapper.decodeWeatherRequest(json:)` と同じように、エンコードエラーを `AppError` で丸め込みたいとしたら、こちらも同様に `try?` とオプショナルバインディングを使って `guard` の中でエラー判定する方法も考えられます。

Reply:
@es-kumagai 
ありがとうございます。
原因不明のエラーが表示されるのは意図しておりません。
guard の理解を間違っていて、try の段階でエラーが発生することを気づきました。
69a0d469e19f4bc3331ee18e835f102ae4e6c5a8
上記のコミットで decode と同じ構成に修正しました。